Failed to run scheduled task ClearExpiredAdminEvents Oracle

### Area

storage

### Describe the bug

A ScheduledTaskRunner task named ClearExpiredAdminEvents periodically runs and fails when Keycloak is configured using the Oracle database.  This is similar to the PostgreSQL issue https://github.com/keycloak/keycloak/issues/14294.

### Version

20.0.1

### Expected behavior

The ScheduledTaskRunner task ClearExpiredAdminEvents should not fail and fill the log with exceptions.

### Actual behavior

Here is the stacktrace log output:

```
2022-11-16 13:36:53,518 WARN  [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (Timer-0) SQL Error: 932, SQLState: 42000
2022-11-16 13:36:53,519 ERROR [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (Timer-0) ORA-00932: inconsistent datatypes: expected - got NCLOB

2022-11-16 13:36:53,527 ERROR [org.keycloak.services.scheduled.ScheduledTaskRunner] (Timer-0) Failed to run scheduled task ClearExpiredAdminEvents: org.hibernate.exception.SQLGrammarException: could not extract ResultSet
        at org.hibernate.exception.internal.SQLExceptionTypeDelegate.convert(SQLExceptionTypeDelegate.java:63)
        at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:37)
        at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
        at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99)
        at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.extract(ResultSetReturnImpl.java:67)
        at org.hibernate.loader.Loader.getResultSet(Loader.java:2322)
        at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2075)
        at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2037)
        at org.hibernate.loader.Loader.scroll(Loader.java:2945)
        at org.hibernate.loader.hql.QueryLoader.scroll(QueryLoader.java:610)
        at org.hibernate.hql.internal.ast.QueryTranslatorImpl.scroll(QueryTranslatorImpl.java:452)
        at org.hibernate.engine.query.spi.HQLQueryPlan.performScroll(HQLQueryPlan.java:353)
        at org.hibernate.internal.SessionImpl.scroll(SessionImpl.java:1601)
        at org.hibernate.query.internal.AbstractProducedQuery.doScroll(AbstractProducedQuery.java:1588)
        at org.hibernate.query.internal.AbstractProducedQuery.scroll(AbstractProducedQuery.java:1574)
        at org.hibernate.query.internal.AbstractProducedQuery.stream(AbstractProducedQuery.java:1598)
        at org.hibernate.query.criteria.internal.compile.CriteriaQueryTypeQueryAdapter.stream(CriteriaQueryTypeQueryAdapter.java:91)
        at org.hibernate.query.Query.getResultStream(Query.java:1140)
        at org.keycloak.events.jpa.JpaEventStoreProvider.clearExpiredAdminEvents(JpaEventStoreProvider.java:292)
        at org.keycloak.events.jpa.JpaEventStoreProviderFactory.invalidate(JpaEventStoreProviderFactory.java:69)
        at org.keycloak.services.DefaultKeycloakSessionFactory.lambda$invalidate$0(DefaultKeycloakSessionFactory.java:368)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
        at java.base/java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1693)
        at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
        at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:274)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1693)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
        at org.keycloak.services.DefaultKeycloakSessionFactory.invalidate(DefaultKeycloakSessionFactory.java:368)
        at org.keycloak.services.DefaultKeycloakSession.invalidate(DefaultKeycloakSession.java:123)
        at org.keycloak.services.scheduled.ClearExpiredAdminEvents.run(ClearExpiredAdminEvents.java:35)
        at org.keycloak.services.scheduled.ClusterAwareScheduledTaskRunner$1.call(ClusterAwareScheduledTaskRunner.java:58)
        at org.keycloak.services.scheduled.ClusterAwareScheduledTaskRunner$1.call(ClusterAwareScheduledTaskRunner.java:54)
        at org.keycloak.cluster.infinispan.InfinispanClusterProvider.executeIfNotExecuted(InfinispanClusterProvider.java:78)
        at org.keycloak.services.scheduled.ClusterAwareScheduledTaskRunner.runTask(ClusterAwareScheduledTaskRunner.java:54)
        at org.keycloak.services.scheduled.ScheduledTaskRunner.run(ScheduledTaskRunner.java:54)
        at org.keycloak.timer.basic.BasicTimerProvider$1.run(BasicTimerProvider.java:53)
        at java.base/java.util.TimerThread.mainLoop(Timer.java:556)
        at java.base/java.util.TimerThread.run(Timer.java:506)
Caused by: java.sql.SQLSyntaxErrorException: ORA-00932: inconsistent datatypes: expected - got NCLOB

        at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:630)
        at oracle.jdbc.driver.T4CTTIoer11.processError(T4CTTIoer11.java:564)
        at oracle.jdbc.driver.T4C8Oall.processError(T4C8Oall.java:1231)
        at oracle.jdbc.driver.T4CTTIfun.receive(T4CTTIfun.java:772)
        at oracle.jdbc.driver.T4CTTIfun.doRPC(T4CTTIfun.java:299)
        at oracle.jdbc.driver.T4C8Oall.doOALL(T4C8Oall.java:512)
        at oracle.jdbc.driver.T4CPreparedStatement.doOall8(T4CPreparedStatement.java:163)
        at oracle.jdbc.driver.T4CPreparedStatement.executeForDescribe(T4CPreparedStatement.java:1010)
        at oracle.jdbc.driver.OracleStatement.prepareDefineBufferAndExecute(OracleStatement.java:1271)
        at oracle.jdbc.driver.OracleStatement.executeMaybeDescribe(OracleStatement.java:1149)
        at oracle.jdbc.driver.OracleStatement.executeSQLSelect(OracleStatement.java:1661)
        at oracle.jdbc.driver.OracleStatement.doExecuteWithTimeout(OracleStatement.java:1470)
        at oracle.jdbc.driver.OraclePreparedStatement.executeInternal(OraclePreparedStatement.java:3761)
        at oracle.jdbc.driver.OraclePreparedStatement.executeQuery(OraclePreparedStatement.java:3936)
        at oracle.jdbc.driver.OraclePreparedStatementWrapper.executeQuery(OraclePreparedStatementWrapper.java:1102)
        at io.agroal.pool.wrapper.PreparedStatementWrapper.executeQuery(PreparedStatementWrapper.java:78)
        at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.extract(ResultSetReturnImpl.java:57)
        ... 41 more
Caused by: Error : 932, Position : 178, Sql = select realmattri0_.NAME as name1_46_, realmattri0_.REALM_ID as realm_id3_46_, realmattri0_.VALUE as value2_46_ from REALM_ATTRIBUTE realmattri0_ where realmattri0_.NAME=:1  and realmattri0_.VALUE<>:2 , OriginalSql = select realmattri0_.NAME as name1_46_, realmattri0_.REALM_ID as realm_id3_46_, realmattri0_.VALUE as value2_46_ from REALM_ATTRIBUTE realmattri0_ where realmattri0_.NAME=? and realmattri0_.VALUE<>?, Error Msg = ORA-00932: inconsistent datatypes: expected - got NCLOB
```

### How to Reproduce?

I have created a Docker Compose reproducer GitHub project.

1. Grab project
```
git clone https://github.com/slominskir/keycloak_oracle
cd keycloak_oracle
```

2. Launch Compose
```
docker compose up
```

3. Login to admin console

http://localhost:8080

Login with username `admin` and password `admin`

4. Enable Admin Scheduled Events

Navigate admin console to `Realm Settings` > `Events` > `Admin event settings`

Click on the `Save events` toggle to enable it.

<img width="960" alt="Screenshot" src="https://user-images.githubusercontent.com/2642119/202205409-0ae79739-9757-4c41-ac31-0ba21901cc24.png">

**Note**: I've actually seen the scheduled task run and cause the exception without enabling Admin events.   This may be only with the ClusterAwareScheduledTaskRunner, see: [keycloak_cluster example](https://github.com/slominskir/keycloak_cluster)

4. Wait for scheduled task to run and observe logs

### Anything else?

_No response_

--- a/model/jpa/src/main/java/org/keycloak/events/jpa/JpaEventStoreProvider.java
+++ b/model/jpa/src/main/java/org/keycloak/events/jpa/JpaEventStoreProvider.java
@@ -35,15 +35,9 @@ import org.keycloak.models.jpa.entities.RealmAttributeEntity;
 import org.keycloak.models.jpa.entities.RealmAttributes;
 import org.keycloak.models.jpa.entities.RealmEntity;
 import org.keycloak.models.utils.KeycloakModelUtils;
-import org.keycloak.provider.InvalidationHandler;
-import org.keycloak.timer.ScheduledTask;
 
 import javax.persistence.EntityManager;
-import javax.persistence.Query;
 import javax.persistence.TypedQuery;
-import javax.persistence.criteria.CriteriaBuilder;
-import javax.persistence.criteria.CriteriaQuery;
-import javax.persistence.criteria.Root;
 import java.io.IOException;
 import java.util.HashMap;
 import java.util.List;
@@ -284,13 +278,10 @@ public class JpaEventStoreProvider implements EventStoreProvider {
     }
 
     protected void clearExpiredAdminEvents() {
-        CriteriaBuilder cb = em.getCriteriaBuilder();
-        CriteriaQuery<RealmAttributeEntity> cr = cb.createQuery(RealmAttributeEntity.class);
-        Root<RealmAttributeEntity> root = cr.from(RealmAttributeEntity.class);
-        // unable to cast the CLOB to a BIGINT in the select for H2 2.x, therefore comparing strings only in the DB, and filtering again in the next statement
-        cr.select(root).where(cb.and(cb.equal(root.get("name"),RealmAttributes.ADMIN_EVENTS_EXPIRATION),cb.notEqual(root.get("value"), "0")));
-        Map<Long, List<RealmAttributeEntity>> realms = em.createQuery(cr).getResultStream()
-                // filtering again on the attribute as paring the CLOB to BIGINT didn't work in H2 2.x
+        TypedQuery<RealmAttributeEntity> query = em.createNamedQuery("selectRealmAttributesNotEmptyByName", RealmAttributeEntity.class)
+                .setParameter("name", RealmAttributes.ADMIN_EVENTS_EXPIRATION);
+        Map<Long, List<RealmAttributeEntity>> realms = query.getResultStream()
+                // filtering again on the attribute as parsing the CLOB to BIGINT didn't work in H2 2.x, and it also different on OracleDB
                 .filter(attribute -> {
                     try {
                         return Long.parseLong(attribute.getValue()) > 0;

--- a/model/jpa/src/main/java/org/keycloak/models/jpa/entities/RealmAttributeEntity.java
+++ b/model/jpa/src/main/java/org/keycloak/models/jpa/entities/RealmAttributeEntity.java
@@ -36,7 +36,8 @@ import java.io.Serializable;
  * @version $Revision: 1 $
  */
 @NamedQueries({
-        @NamedQuery(name="deleteRealmAttributesByRealm", query="delete from RealmAttributeEntity attr where attr.realm = :realm")
+        @NamedQuery(name="deleteRealmAttributesByRealm", query="delete from RealmAttributeEntity attr where attr.realm = :realm"),
+        @NamedQuery(name="selectRealmAttributesNotEmptyByName", query="select ra from RealmAttributeEntity ra WHERE ra.name = :name and length(ra.value) > 0")
 })
 @Table(name="REALM_ATTRIBUTE")
 @Entity

--- a/testsuite/integration-arquillian/pom.xml
+++ b/testsuite/integration-arquillian/pom.xml
@@ -469,6 +469,7 @@
         <profile>
             <id>db-allocator-db-mysql</id>
             <properties>
+                <keycloak.storage.connections.vendor>mysql</keycloak.storage.connections.vendor>
                 <dballocator.type>mysql80</dballocator.type>
                 <dballocator.skip>false</dballocator.skip>
             </properties>
@@ -502,6 +503,7 @@
         <profile>
             <id>db-allocator-db-postgres</id>
             <properties>
+                <keycloak.storage.connections.vendor>postgres</keycloak.storage.connections.vendor>
                 <dballocator.type>postgresql132</dballocator.type>
                 <dballocator.skip>false</dballocator.skip>
             </properties>
@@ -509,6 +511,7 @@
         <profile>
             <id>db-allocator-db-postgresplus</id>
             <properties>
+                <keycloak.storage.connections.vendor>postgres</keycloak.storage.connections.vendor>
                 <dballocator.type>postgresplus131</dballocator.type>
                 <dballocator.skip>false</dballocator.skip>
             </properties>
@@ -543,6 +546,7 @@
         <profile>
             <id>db-allocator-db-mariadb</id>
             <properties>
+                <keycloak.storage.connections.vendor>mariadb</keycloak.storage.connections.vendor>
                 <dballocator.type>mariadb_galera_103</dballocator.type>
                 <dballocator.skip>false</dballocator.skip>
             </properties>
@@ -577,6 +581,7 @@
         <profile>
             <id>db-allocator-db-mssql2019</id>
             <properties>
+                <keycloak.storage.connections.vendor>mssql</keycloak.storage.connections.vendor>
                 <dballocator.type>mssql2019</dballocator.type>
                 <dballocator.skip>false</dballocator.skip>
             </properties>
@@ -613,6 +618,7 @@
         <profile>
             <id>db-allocator-db-oracleRAC</id>
             <properties>
+                <keycloak.storage.connections.vendor>oracle</keycloak.storage.connections.vendor>
                 <dballocator.type>oracle19cRAC</dballocator.type>
                 <dballocator.skip>false</dballocator.skip>
             </properties>

